
With a dataset of fish species, with some of it characteristic like it vertical, diagonal, length, height, and width. We will try to predict the weight of the fish based on their characteristic. We will use Linear Regression Method to see whether the weight of the fish related to their characteristic.

- Species: Species name of fish
- Weight: Weight of fish in gram
- Length1: Vertical length in cm
- Length2: Diagonal length in cm
- Length3: Cross length in cm
- Height: Height in cm
-Width: Diagonal width in cm

In [1]:
#import required libraries
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('https://github.com/ybifoundation/Dataset/raw/main/Fish.csv')

In [3]:
df

Category Species  Weight   Height   Width  Length1  Length2  Length3
0           1   Bream   242.0  11.5200  4.0200     23.2     25.4     30.0
1           1   Bream   290.0  12.4800  4.3056     24.0     26.3     31.2
2           1   Bream   340.0  12.3778  4.6961     23.9     26.5     31.1
3           1   Bream   363.0  12.7300  4.4555     26.3     29.0     33.5
4           1   Bream   430.0  12.4440  5.1340     26.5     29.0     34.0
..        ...     ...     ...      ...     ...      ...      ...      ...
154         6   Smelt    12.2   2.0904  1.3936     11.5     12.2     13.4
155         6   Smelt    13.4   2.4300  1.2690     11.7     12.4     13.5
156         6   Smelt    12.2   2.2770  1.2558     12.1     13.0     13.8
157         6   Smelt    19.7   2.8728  2.0672     13.2     14.3     15.2
158         6   Smelt    19.9   2.9322  1.8792     13.8     15.0     16.2

[159 rows x 8 columns]

In [4]:
df.head()

Category Species  Weight   Height   Width  Length1  Length2  Length3
0         1   Bream   242.0  11.5200  4.0200     23.2     25.4     30.0
1         1   Bream   290.0  12.4800  4.3056     24.0     26.3     31.2
2         1   Bream   340.0  12.3778  4.6961     23.9     26.5     31.1
3         1   Bream   363.0  12.7300  4.4555     26.3     29.0     33.5
4         1   Bream   430.0  12.4440  5.1340     26.5     29.0     34.0

In [5]:
df.tail()

Category Species  Weight  Height   Width  Length1  Length2  Length3
154         6   Smelt    12.2  2.0904  1.3936     11.5     12.2     13.4
155         6   Smelt    13.4  2.4300  1.2690     11.7     12.4     13.5
156         6   Smelt    12.2  2.2770  1.2558     12.1     13.0     13.8
157         6   Smelt    19.7  2.8728  2.0672     13.2     14.3     15.2
158         6   Smelt    19.9  2.9322  1.8792     13.8     15.0     16.2

In [6]:
df.describe()

Category       Weight      Height       Width     Length1  \
count  159.000000   159.000000  159.000000  159.000000  159.000000   
mean     3.264151   398.326415    8.970994    4.417486   26.247170   
std      1.704249   357.978317    4.286208    1.685804    9.996441   
min      1.000000     0.000000    1.728400    1.047600    7.500000   
25%      2.000000   120.000000    5.944800    3.385650   19.050000   
50%      3.000000   273.000000    7.786000    4.248500   25.200000   
75%      4.500000   650.000000   12.365900    5.584500   32.700000   
max      7.000000  1650.000000   18.957000    8.142000   59.000000   

          Length2     Length3  
count  159.000000  159.000000  
mean    28.415723   31.227044  
std     10.716328   11.610246  
min      8.400000    8.800000  
25%     21.000000   23.150000  
50%     27.300000   29.400000  
75%     35.500000   39.650000  
max     63.400000   68.000000

In [7]:
df.describe(include='all')

Category Species       Weight      Height       Width     Length1  \
count   159.000000     159   159.000000  159.000000  159.000000  159.000000   
unique         NaN       7          NaN         NaN         NaN         NaN   
top            NaN   Perch          NaN         NaN         NaN         NaN   
freq           NaN      56          NaN         NaN         NaN         NaN   
mean      3.264151     NaN   398.326415    8.970994    4.417486   26.247170   
std       1.704249     NaN   357.978317    4.286208    1.685804    9.996441   
min       1.000000     NaN     0.000000    1.728400    1.047600    7.500000   
25%       2.000000     NaN   120.000000    5.944800    3.385650   19.050000   
50%       3.000000     NaN   273.000000    7.786000    4.248500   25.200000   
75%       4.500000     NaN   650.000000   12.365900    5.584500   32.700000   
max       7.000000     NaN  1650.000000   18.957000    8.142000   59.000000   

           Length2     Length3  
count   159.000000  159.000000  
unique         NaN         NaN  
top            NaN         NaN  
freq           NaN         NaN  
mean     28.415723   31.227044  
std      10.716328   11.610246  
min       8.400000    8.800000  
25%      21.000000   23.150000  
50%      27.300000   29.400000  
75%      35.500000   39.650000  
max      63.400000   68.000000

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159 entries, 0 to 158
Data columns (total 8 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Category  159 non-null    int64  
 1   Species   159 non-null    object 
 2   Weight    159 non-null    float64
 3   Height    159 non-null    float64
 4   Width     159 non-null    float64
 5   Length1   159 non-null    float64
 6   Length2   159 non-null    float64
 7   Length3   159 non-null    float64
dtypes: float64(6), int64(1), object(1)
memory usage: 10.1+ KB


In [9]:
df.isna().sum()

Category    0
Species     0
Weight      0
Height      0
Width       0
Length1     0
Length2     0
Length3     0
dtype: int64

In [22]:
#let's split the data

Y = df['Weight']

X = df[['Height','Width','Length1','Length2','Length3']]

In [23]:
Y.shape

(159,)

In [24]:
X.shape

(159, 5)

In [25]:
from sklearn.model_selection import train_test_split

In [26]:
X_train,X_test,Y_train,Y_test = train_test_split(X,Y,random_state=42)

In [27]:
X_train.shape

(119, 5)

In [28]:
X_test.shape

(40, 5)

In [29]:
Y_train.shape

(119,)

In [30]:
Y_test.shape

(40,)

As the output or the target feature is a continuous one we opt for regression model.

In [31]:
from sklearn.linear_model import LinearRegression

In [32]:
LR = LinearRegression()

In [33]:
LR.fit(X_train,Y_train)

LinearRegression()

In [34]:
LR.intercept_

-513.6495118585174

In [35]:
LR.coef_

array([ 25.8537637 ,  17.27336678,  51.43500443,   6.44757662,
       -29.95777296])

In [36]:
y_pred = LR.predict(X_test)

In [37]:
y_pred

array([  78.2179719 , -151.59539671,  397.38857993,  298.10073482,
        178.36549965,  775.94818257, -208.79944173,  277.55871032,
        268.5394453 , 1130.48977796,  610.60076919,  791.72220927,
        585.94870664,  177.44207964,  655.92140246,  788.9223388 ,
        922.95028288,  361.16934648,  261.5994571 ,  590.85586966,
       -169.38618777,  628.16580372,  526.92201633,  519.19402524,
        792.88881922,  914.66443363,  221.35246835,  303.00727312,
       -232.61156442, -171.42271772,  708.16788159,   21.2757673 ,
         11.2400173 ,  831.56148369,  355.95130937,  169.71509125,
        175.73620166,  124.57853959,  227.66047592,  831.4504363 ])

In [38]:
#accuracy metrics
from sklearn.metrics import mean_absolute_error,mean_absolute_percentage_error,r2_score,mean_squared_error

In [39]:
print(mean_absolute_error(Y_test,y_pred))

95.36828906864321


In [40]:
print(mean_absolute_percentage_error(Y_test,y_pred))

3.0954777787243133


In [41]:
print(r2_score(Y_test,y_pred))

0.8828776020785211


Therefore this model predicts 88 % of the values correctly.